governance/status.go

```diff
@@ -0,0 +1,413 @@
+			found := false
+			for _, item2 := range a2 {
+				if item1.IsSame(item2) {
+					found = true
+				}
+			}
+			if !found {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// Report the containers status and connectivity status to the exchange.
+func (w *GovernanceWorker) reportStatus() {
+
+	go func() {
+
+		// This is the amount of time for the governance routine to wait.
+		waitTime := w.Config.Edge.ReportDeviceStatusInterval
+		if waitTime == 0 { // never report status when the interval is set to 0
+			glog.V(5).Info("ReportDeviceStatusInterval is 0. The status report to the exchange is turned off.")
+			return
+		}
+
+		for {
+			glog.Info("started the status report to the exchange.")
+
+			var device_status DeviceStatus
+
+			// get connectivity
+			connect := make(map[string]bool, 0)
+			for _, host := range HORIZON_SERVERS {
+				if err := cutil.CheckConnectivity(host); err != nil {
+					glog.Errorf(logString(fmt.Sprintf("Error checking connectivity for %s: %v", host, err)))
+					connect[host] = false
+				} else {
+					connect[host] = true
+				}
+			}
+			device_status.Connectivity = connect
+
+			// get microservice status
+			if ms_status, err := w.getMicroserviceStatus(); err != nil {
+				glog.Errorf(logString(fmt.Sprintf("Error getting microservice container status: %v", err)))
+			} else {
+				device_status.Microservices = ms_status
+			}
+
+			// get workload status
+			if wl_status, err := w.getWorkloadStatus(); err != nil {
+				glog.Errorf(logString(fmt.Sprintf("Error getting microservice container status: %v", err)))
+			} else {
+				device_status.Workloads = wl_status
+			}
+
+			//compare the device status with the one gotten from last time
+			if jbytes, err := json.Marshal(&device_status); err != nil {
+				glog.V(5).Infof(logString(fmt.Sprintf("Failed to convert the device status %v to json: %v", device_status, err)))
+			} else if Device_Status != nil && device_status.IsSame(*Device_Status) {
+				glog.V(5).Infof(logString(fmt.Sprintf("device status is the same as last time. No need to report: %v", string(jbytes[:]))))
+			} else {
+
+				Device_Status = &device_status
+				glog.V(5).Infof(logString(fmt.Sprintf("device status to report to the exchange: %v", string(jbytes[:]))))
+
+				// report the status to the exchange
+				if err := w.writeStatusToExchange(&device_status); err != nil {
+					glog.Errorf(logString(err))
+				}
+			}
+
+			// Sleep
+			time.Sleep(time.Duration(waitTime) * time.Second)
+		}
+	}()
+}
+
+// Find the status for all the microservices
+func (w *GovernanceWorker) getMicroserviceStatus() ([]MicroserviceStatus, error) {
+
+	// Filter to return all instances for a msdef
+	msdefFilter := func(msdef_id string) persistence.MIFilter {
+		return func(a persistence.MicroserviceInstance) bool {
+			return a.MicroserviceDefId == msdef_id
+		}
+	}
+
+	status := make([]MicroserviceStatus, 0)
+
+	if client, err := docker.NewClient(w.Config.Edge.DockerEndpoint); err != nil {
```

Comment:
If you factor this docker call out of this API, then you can (a) only call it once, and (b) I think getMicroserviceStatus and getWorkloadStatus and getContainerStatus will become unittestable functions. It is hard to mock the docker call so you can't UT these large functions.

Reply:
I do not understand the comment. Need to discuss this with you.